.specs/cojson-crypto-rust-migration/design.md

```diff
@@ -0,0 +1,138 @@
+## Overview
+
+This feature removes TypeScript-side canonicalization (`stableStringify`) from the crypto pipeline by moving **deterministic serialization of predefined data structures** into Rust, and then routing all providers (`WasmCrypto`, `NapiCrypto`, `RNCrypto`) through those Rust implementations.
+
+Key goals:
+
+- Preserve the current TypeScript API surface and string formats (`signerSecret_z…`, `signature_z…`, `encrypted_U…`, `sealed_U…`, `hash_z…`, `shortHash_z…`).
+- Preserve backwards compatibility for signatures, sealed payloads, encrypted payloads, and hashes by matching `stableStringify` semantics in Rust.
+- Do **not** change random number generation behavior (US7).
+
+Non-goals:
+
+- Changing the randomness sources, RNG libraries, or key generation formats.
+- Introducing a new externally-visible cryptographic scheme or changing existing prefixes/encodings.
+
+## Architecture / components
+
+### 1) Rust-owned data models for all crypto inputs
+
+Instead of implementing a “`stableStringify`-equivalent” function that takes arbitrary JSON text, we treat determinism as a **data modeling** problem:
+
+- Every value that currently flows into `stableStringify(...)` is represented in Rust as a concrete `struct`/`enum`.
+- Deterministic encoding is achieved by `serde` serialization of those Rust types (plus deterministic map types where needed), not by a “canonicalize JSON string” utility.
+
+This matches your constraint: **Rust builds/owns the structures**, and serialization happens from those structures.
```

Comment:
We found that `stableStringify` isn't really required everywhere it's used.

So we can also remove it's usage from where it's not required. (e.g. encrypted changes and meta)